### ISSUE TITLE
Add foldMap for IntSet

### ISSUE THREAD

**Issue by Fuuzetsu** [haskell/containers#584]
We're looking for something like

```haskell
foldMap :: Monoid m => (Key -> m) -> IntSet -> m
```

This in turn allows other functions such as `mapM_` to be implemented:

```haskell
newtype IO_ = IO_ { _unIO :: IO () }
instance Monoid IO_ where …
unIO_ $ foldMap (IO_ . f) 
```

Currently implementing either is not possible without access to internals. As a result, we often end up running code like `mapM_ (IntSet.toList s) f` which is slower.

I think implementation is fairly straight forward in `IntSet.Internal`, here's one I came up with just by copying and pasting unexported stuff blindly out of the module

```haskell
foldMapBits :: Monoid m => Int -> (Int -> m) -> Nat -> m
foldMapBits prefix f bitmap = go bitmap
  where go 0 = mempty
        go bm = f (prefix + bi) <> go (bm `xor` bitmask)
          where !bitmask = lowestBitMask bm
                !bi = indexOfTheOnlyBit bitmask

foldMapSet :: Monoid m => (Int -> m) -> I.IntSet -> m
foldMapSet f s =
    let go (I.Bin _ _ s1 s2) = go s1 <> go s2
        go (I.Tip kx bm) = foldMapBits kx f bm
        go I.Nil = mempty
    in go s
```

This seems to run about ~25-30% faster than going through `fromList` in my rough benchmarks so seems like a very worth-while addition.

**Comment by treeowl** [comment]
Yes, we should definitely offer this. Can you put together a pull request? Please be sure to add QuickCheck properties. I believe the most appropriate monoid for the tests is `Data.Sequence.Seq A`. Anything in `IntSet` should be in `Set` as well, so let's make it so. I would also appreciate if you'd send an official proposal to libraries@haskell.org, just in case people have Strong Opinions about the name. I think your choice is correct, but it's always good to hear from the community.
